File: src/с/bsk_stack.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct {
    char color; // 'W', 'B', 'R'
    int value;
} Node;

typedef struct {
    Node *data;
    size_t cap;
    size_t n;
    double whiteRatio;
    double blueRatio;
} BSKStack;

BSKStack* bsk_new(double w, double b) {
    BSKStack *s = malloc(sizeof(BSKStack));
    s->cap = 16; s->n = 0;
    s->data = malloc(sizeof(Node)*s->cap);
    s->whiteRatio = w; s->blueRatio = b;
    return s;
}
void ensure_cap(BSKStack *s) {
    if(s->n >= s->cap) {
        s->cap *= 2;
        s->data = realloc(s->data, sizeof(Node)*s->cap);
    }
}
void bsk_push(BSKStack *s, int v) {
    ensure_cap(s);
    // insert at front
    memmove(&s->data[1], &s->data[0], sizeof(Node)*s->n);
    s->data[0].color = 'W';
    s->data[0].value = v;
    s->n++;
}
void bsk_zones(BSKStack *s, size_t *w, size_t *b) {
    size_t n = s->n;
    *w = (size_t)(n * s->whiteRatio);
    *b = (size_t)(n * (s->whiteRatio + s->blueRatio));
}
int bsk_pop(BSKStack *s, int *out) {
    if(s->n == 0) return 0;
    size_t w,b; bsk_zones(s,&w,&b);
    for(size_t i=0;i<w && i<s->n;i++){
        if(s->data[i].color=='W' || s->data[i].color=='B') {
            *out = s->data[i].value;
            memmove(&s->data[i], &s->data[i+1], sizeof(Node)*(s->n-i-1));
            s->n--; return 1;
        }
    }
    for(size_t i=w;i<b && i<s->n;i++){
        if(s->data[i].color=='B') {
            *out = s->data[i].value;
            memmove(&s->data[i], &s->data[i+1], sizeof(Node)*(s->n-i-1));
            s->n--; return 1;
        }
    }
    // red locked
    return -1;
}
void bsk_update(BSKStack *s) {
    size_t w,b; bsk_zones(s,&w,&b);
    for(size_t i=0;i<s->n;i++){
        if(i < w) s->data[i].color='W';
        else if(i < b) s->data[i].color='B';
        else s->data[i].color='R';
    }
}
size_t bsk_len(BSKStack *s){ return s->n;}
void bsk_free(BSKStack *s){ free(s->data); free(s); }

File: src/go/bsk_stack.go
! bsk_stack.f90
module bsk_mod
  implicit none
  private
  public :: BSKType, bsk_new, bsk_push, bsk_pop, bsk_update, bsk_size

  type :: Node
    character(len=1) :: color
    integer :: value
  end type Node

  type :: BSKType
    type(Node), allocatable :: data(:)
    real :: whiteRatio = 0.3
    real :: blueRatio = 0.4
  end type BSKType

contains

  function bsk_new() result(s)
    type(BSKType) :: s
    allocate(s%data(0))
  end function

  subroutine bsk_push(s, v)
    type(BSKType), intent(inout) :: s
    integer, intent(in) :: v
    integer :: n
    n = size(s%data)
    call move_alloc(s%data, s%data, stat=)
    s%data = [ (Node('W', v)) , s%data ]
  end subroutine

  integer function bsk_pop(s)
    type(BSKType), intent(inout) :: s
    integer :: n, w, b, i
    n = size(s%data)
    if (n==0) then
      stop "empty"
    end if
    w = int(n * s%whiteRatio)
    b = int(n * (s%whiteRatio + s%blueRatio))
    do i=1,w
      if (s%data(i)%color == 'W' .or. s%data(i)%color == 'B') then
        bsk_pop = s%data(i)%value
        s%data = [ s%data(1:i-1), s%data(i+1:)/ ]
        return
      end if
    end do
    do i=w+1,b
      if (s%data(i)%color == 'B') then
        bsk_pop = s%data(i)%value
        s%data = [ s%data(1:i-1), s%data(i+1:)/ ]
        return
      end if
    end do
    stop "Red zone locked"
  end function

  subroutine bsk_update(s)
    type(BSKType), intent(inout) :: s
    integer :: n, w, b, i
    n = size(s%data)
    w = int(n * s%whiteRatio)
    b = int(n * (s%whiteRatio + s%blueRatio))
    do i=1,n
      if (i <= w) s%data(i)%color = 'W'
      else if (i <= b) s%data(i)%color = 'B'
      else s%data(i)%color = 'R'
    end do
  end subroutine

  integer function bsk_size(s)
    type(BSKType), intent(in) :: s
    bsk_size = size(s%data)
  end function

end module
